build-system/tasks/e2e/selenium-webdriver-controller.js

```diff
@@ -127,13 +171,34 @@ class SeleniumWebDriverController {
    * @override
    */
   async findElementsXPath(xpath) {
-    const byXpath = By.xpath(xpath);
-
-    await this.driver.wait(until.elementLocated(byXpath));
-    const webElements = await this.driver.findElements(byXpath);
+    await this.maybeInstallXpath_();
+
+    const webElements = await this.driver.wait(new Condition('', async() => {
+      try {
+        const root = await this.getRoot_();
+        const results = await this.evaluate(queryXpath, xpath, root);
+        return results;
+      } catch (nse) {
+        return null;
+      }
+    }), ELEMENT_WAIT_TIMEOUT);
     return webElements.map(webElement => new ElementHandle(webElement, this));
   }
 
+  /**
+   * Install a third-party XPath library if it is not already installed.
+   * @return {!Promise}
+   */
+  async maybeInstallXpath_() {
+    if (this.isXpathInstalled_) {
+      return;
+    }
+    this.isXpathInstalled_ = true;
+
+    const script = fs.readFileSync('third_party/wgxpath/wgxpath.js', 'utf8');
```

Comment:
I'm pretty sure that fs.readFileAsync returns a `Promise`, so would be better to do that rather than have a sync call here when everything else is async.

Reply:
Nice, TIL. I thought there were only continuation passing `readFile` and the sync `readFileSync`

Reply:
Looks like there is no `readFileAsync`, but there is an [experimental feature fsPromises](https://nodejs.org/dist/latest-v10.x/docs/api/fs.html#fs_fspromises_readfile_path_options):
```js
const {readFile} = require('fs').promises;
```

Reply:
Ah, it looks like we have this in our build-system:

```javascript
const BBPromise = require('bluebird');
const fs = BBPromise.promisifyAll(require('fs'));
```

which is what is adding the async versions. Could use that to stay consistent.

Reply:
Huh neat, thanks